File: src/main/java/me/andre111/d20server/scripting/variable/parser/VariableParserParent.java
package me.andre111.d20server.scripting.variable.parser;

import java.util.HashMap;
import java.util.Map;

import me.andre111.d20server.scripting.ParserContext;
import me.andre111.d20server.scripting.ScriptException;
import me.andre111.d20server.scripting.variable.Variable;

public class VariableParserParent extends VariableParser {
	private Map<String, VariableParser> children = new HashMap<>();
	
	public VariableParserParent addChild(String name, VariableParser child) {
		children.put(name, child);
		return this;
	}

	@Override
	public Variable parse(ParserContext context, String fullName, String name) throws ScriptException{
		VariableParser child = children.get(getChildParserName(name));
		if(child == null) {
			throw new ScriptException("Unknown variable "+fullName);
		}
		
		updateContext(context, name);
		
		return child.parse(context, fullName, getChildString(name));
	}

	// default implementation for basic parents "<childName>.<childString>"
	protected String getChildParserName(String name) {
		String[] split = name.split("\\.", 2);
		return split[0];
	}
	protected String getChildString(String name) {
		String[] split = name.split("\\.", 2);
		return split.length==2 ? split[1] : "";
	}
	protected void updateContext(ParserContext context, String name) {
	}
}

File: core/common/scripting/variable/parser/variable-parser-parent.js
import { VariableParser } from './variable-parser.js';

export class VariableParserParent extends VariableParser {
    children = {};

    constructor() {
        super();
    }

    addChild(name, child) {
        if(!(child instanceof VariableParser)) throw new Error('Can only add instances of VariableParser');
        this.children[name] = child;
        return this;
    }

    parse(context, fullName, name) {
        const childName = this.getChildParserName(name);
        const child = this.children[childName];
        if(!child) throw new Error(`Unknown variable ${fullName}, no child with name ${childName}`);

        this.updateContext(context, name);

        return child.parse(context, fullName, this.getChildString(name));
    }

    updateContext(context, name) {
    }

    // default implementation for basic parents '<childName>.<childString>'
    getChildParserName(name) {
        const split = name.split('.');
        return split[0];
    }

    getChildString(name) {
        const split = name.split('.');
        return split.length >= 2 ? split.slice(1).join('.') : '';
    }
}
